lib/src/package_config_impl.dart

```diff
@@ -179,3 +193,107 @@ void _validatePackageData(
         languageVersion, "languageVersion", "Invalid language version format");
   }
 }
+
+abstract class PackageTree {
+  SimplePackage /*?*/ packageOf(Uri file);
+}
+
+/// Packages of a package configuration ordered by root path.
+///
+/// A package is said to be inside another package if the root path URI of
+/// the latter is a prefix of the root path URI of the former.
+/// No two packages of a package may have the same root path, so this
+/// path prefix ordering defines a tree-like partial ordering on packages
+/// of a configuration.
+///
+/// The package tree contains an ordered mapping of unrelated packages
+/// (represented by their name) to their immediately nested packages' names.
+class MutablePackageTree implements PackageTree {
+  final List<SimplePackage> packages = [];
+  Map<String, MutablePackageTree /*?*/ > /*?*/ _packageChildren;
+
+  /// Tries to (add) `package` to the tree.
+  ///
+  /// If another package is found with the *same* path, adding fails
+  /// and that package is returned. Returns `null` on success.
+  SimplePackage /*?*/ add(int start, SimplePackage package) {
```

Comment:
nit: Maybe rename this `addOrReturnConflictingPackage(SimplePackage package, [int start = 0])`... One could also throw in the case of conflicts, as speed is unimportant in the case of a conflict.

Reply:
That's not a bad idea. Let's throw.